extensions/notebook/src/book/bookTrustManager.ts

```diff
@@ -0,0 +1,161 @@
+ *--------------------------------------------------------------------------------------------*/
+import * as path from 'path';
+import * as vscode from 'vscode';
+import * as constants from './../common/constants';
+import { BookTreeItem } from './bookTreeItem';
+import { BookModel } from './bookModel';
+
+export interface IBookTrustManager {
+	isNotebookTrustedByDefault(notebookUri: string): boolean;
+	setBookAsTrusted(bookRootPath: string): boolean;
+	setBookAsUntrusted(bookRootPath: string): boolean;
+}
+
+export interface IBookTrustManagerWorkspaceDetails {
+	workspaceFolders: vscode.WorkspaceFolder[];
+	getConfiguration(section?: string | undefined): vscode.WorkspaceConfiguration;
+}
+
+enum TrustBookOperation {
+	Add,
+	Remove
+}
+
+export class BookTrustManager implements IBookTrustManager {
+
+	private static notebookConfiguration: string = constants.notebookConfigKey;
+	private static notebookTrustedBooksConfiguration: string = constants.trustedBooksConfigKey;
+	private trustedLocalBooks: string[] = [];
+
+	constructor(private books: BookModel[], private workspaceDetails?: IBookTrustManagerWorkspaceDetails) {
+		if (!workspaceDetails) {
+			this.workspaceDetails = {
+				get getConfiguration() {
+					return vscode.workspace.getConfiguration;
+				},
+				get workspaceFolders() {
+					return [].concat(vscode.workspace.workspaceFolders || []);
+				}
+			};
+		}
+	}
+
+	isNotebookTrustedByDefault(notebookUri: string): boolean {
+		let allBooks: BookTreeItem[] = this.getAllBooks();
+		let trustableBookPaths = this.getTrustableBookPaths();
+		let trustableBooks: BookTreeItem[] = allBooks
+			.filter(bookItem => trustableBookPaths.some(trustableBookPath => trustableBookPath === path.join(bookItem.book.root, path.sep)));
+		let isNotebookTrusted = this.isNotebookPartOfBooks(notebookUri, trustableBooks);
+
+		return isNotebookTrusted;
+	}
+
+	isNotebookPartOfBooks(notebookUri: string, books: BookTreeItem[]) {
+		let normalizedNotebookUri = path.normalize(notebookUri);
+		let isPartOfNotebook: boolean = false;
+		let trustedBookItem: BookTreeItem = books.find(bookItem => normalizedNotebookUri.startsWith(path.join(bookItem.book.root, path.sep)));
+
+		if (trustedBookItem) {
+			let fullBookBaseUriWithContent = path.join(trustedBookItem.book.root, 'content', path.sep);
+			let requestingNotebookFormattedUri = normalizedNotebookUri.replace(fullBookBaseUriWithContent, '').replace('.ipynb', '');
+			let notebookInTOC = trustedBookItem.book.tableOfContents.sections.find(jupyterSection => {
+				let normalizedJupyterSectionUrl = jupyterSection.url && path.normalize(jupyterSection.url).replace(/^\\/, '');
+				return normalizedJupyterSectionUrl === requestingNotebookFormattedUri;
+			});
+			isPartOfNotebook = !!notebookInTOC;
+		}
+
+		return isPartOfNotebook;
+	}
+
+	getTrustableBookPaths() {
+		let trustablePaths: string[];
+		let bookPathsInConfig: string[] = this.getTrustedBookPathsInConfig();
+
+		if (this.hasWorkspaceFolders()) {
```

Comment:
So why do you need to check workspace folders here anyways? I thought that we always added books contained in a workspace to the list of books when we opened that workspace - is this not the case?

Reply:
Not exactly. A workspace can contain multiple books. Trusting one book in the workspaces does not extend that same trust to other books in the same workspace. The `getTrustedBookPathsInConfig` retrieves the trusted books in a particular workspace. In our team, we have 4 books in one workspace.

Reply:
Yeah - so what I'm saying is that we already parse the workspaces for existing books when initializing the tree : https://github.com/microsoft/azuredatastudio/blob/master/extensions/notebook/src/book/bookTreeView.ts#L50

I don't think this trust logic needs to know about workspaces at all. It should just keep a list of "trusted books" and then when a book is opened if it's in the list of trusted books it goes and trusts all the notebooks it contains.

Does that make sense?

Reply:
Ah yes, I see what you are saying. The aim here was to store in the workspace configuration, the paths of the books, relative to the workspace, so that trusting relationship can survive relocation. This is specifically needed in our scenario in that we want users to save the trust relationship on file, and share it with other team members. For example, the workspace configuration may look like this:
```
{
	"folders": [
		{
			"path": "."
		}
	],
	"settings": {
		"notebook.trustedBooks": [
			"Books/BookA",
			"Books/BookB"
		]
	}
}
```
Where the `path` property is all the files where the workspace file lives. The model, currently does not have a unique identifier other than the root folder, but that's an absolute path. To opt to not store full paths in configuration, workspace info is pulled into the trust manager to workout the relative paths so that it is cleaner. So if the workspace folder is located at `C:\temp\workspace` (books would be located at `C:\temp\workspace\Books\BookA` and `C:\temp\workspace\Books\BookB`), then saving it as a relative path allows us to maintain the trust relationship, say for when the workpace file is relocated to `D:\newfolder\`.

You are right, if we were to only to make a book trust on a user to user basis, but for extending this trust for other people, the above helps. 

Perhaps there is another avenue for expressing this in the code base. Approached it this by encapsulating the logic in this module for now, and refactor later as the need for it became broader. 

Let me know what you think.

Reply:
Ah, got it. Yeah that's an interesting use case. Definitely behavior I think we'll want although I'm not sure this is the best way. Let me think on it and we can sync up next week - sound good?

Reply:
Ok - we can discuss this more later but I looked at this and generally think what you're doing is fine. But I did notice that it looks like this code is set up to assume that all trusted books are in workspaces. I think this should be expanded to allow trusting a book that was opened directly as well. We'll need to discuss how we'd want to determine whether a book should be saved as a relative path or an absolute path though.

Reply:
Synced in our meeting. Workspace leverage relative path, folders absolute path.

Reply:
# resolved